sweepbatcher/presigned.go

```diff
@@ -0,0 +1,441 @@
+			"for feeRate %v: %w", tx.TxHash(), feeRate, err)
+	}
+
+	// Check the SignTx worked correctly.
+	err = CheckSignedTx(tx, signedTx, batchAmt, feeRate)
+	if err != nil {
+		return fmt.Errorf("signed tx doesn't correspond the "+
+			"unsigned tx: %w", err)
+	}
+
+	return nil
+}
+
+// presign tries to presign batch sweep transactions composed of this batch and
+// the sweep. It signs multiple versions of the transaction to make sure there
+// is a transaction to be published if minRelayFee grows.
+func (b *batch) presign(ctx context.Context, newSweep *sweep) error {
+	if b.cfg.presignedHelper == nil {
+		return fmt.Errorf("presignedHelper is not installed")
+	}
+	if len(b.sweeps) == 0 {
+		return fmt.Errorf("presigning is done when adding to a " +
+			"non-empty batch")
+	}
+
+	// Find the feerate needed to get into next block. Use conf_target=2,
+	nextBlockFeerate, err := b.wallet.EstimateFeeRate(ctx, 2)
+	if err != nil {
+		return fmt.Errorf("failed to get nextBlockFeerate: %w", err)
+	}
+
+	b.log().Infof("nextBlockFeerate is %v", nextBlockFeerate)
+
+	// Create the list of sweeps of the future batch.
+	sweeps := make([]sweep, 0, len(b.sweeps)+1)
+	for _, sweep := range b.sweeps {
+		sweeps = append(sweeps, sweep)
+	}
+	existingSweeps := sweeps
+	sweeps = append(sweeps, *newSweep)
+
+	// Cache the destination address.
+	destAddr, err := b.getSweepsDestAddr(ctx, existingSweeps)
+	if err != nil {
+		return fmt.Errorf("failed to find destination address: %w", err)
+	}
+
+	return presign(
+		ctx, b.cfg.presignedHelper, destAddr, sweeps, nextBlockFeerate,
+	)
+}
+
+// presigner tries to presign a batch transaction.
+type presigner interface {
+	// Presign tries to presign a batch transaction. If the method returns
+	// nil, it is guaranteed that future calls to SignTx on this set of
+	// sweeps return valid signed transactions.
+	Presign(ctx context.Context, tx *wire.MsgTx,
+		inputAmt btcutil.Amount) error
+}
+
+// presign tries to presign batch sweep transactions of the sweeps. It signs
+// multiple versions of the transaction to make sure there is a transaction to
+// be published if minRelayFee grows. If feerate is high, then a presigned tx
+// gets LockTime equal to timeout minus 50 blocks, as a precautionary measure.
+// A feerate is considered high if it is at least 100 sat/vbyte AND is at least
+// 10x of the current next block feerate.
+func presign(ctx context.Context, presigner presigner, destAddr btcutil.Address,
+	sweeps []sweep, nextBlockFeerate chainfee.SatPerKWeight) error {
+
+	if presigner == nil {
+		return fmt.Errorf("presigner is not installed")
+	}
+
+	if len(sweeps) == 0 {
+		return fmt.Errorf("there are no sweeps")
+	}
+
+	if nextBlockFeerate == 0 {
+		return fmt.Errorf("nextBlockFeerate is not set")
+	}
+
+	// Keep track of the total amount this batch is sweeping back.
+	batchAmt := btcutil.Amount(0)
+	for _, sweep := range sweeps {
+		batchAmt += sweep.value
+	}
+
+	// Find when at least one sweep expires.
```

Comment:
nit: "find the sweep with the earliest expiry"

Reply:
Fixed